src/main/scala/is/hail/io/bgen/BgenRecord.scala

```diff
@@ -0,0 +1,176 @@
+}
+
+class BgenRecordV11(compressed: Boolean,
+  nSamples: Int,
+  tolerance: Double) extends BgenRecord {
+
+  override def getValue: Iterable[Genotype] = {
+    require(input != null, "called getValue before serialized value was set")
+
+    val bytes = if (compressed) decompress(input, nSamples * 6) else input
+
+    assert(bytes.length == nSamples * 6)
+
+    resetWarnings()
+
+    val lowerTol = (32768 * (1.0 - tolerance) + 0.5).toInt
+    val upperTol = (32768 * (1.0 + tolerance) + 0.5).toInt
+    assert(lowerTol > 0)
+
+    val noCall: Genotype = new DosageGenotype(-1, null)
+
+    new Iterable[Genotype] {
+      def iterator = new Iterator[Genotype] {
+        var i = 0
+
+        def hasNext: Boolean = i < bytes.length
+
+        def next(): Genotype = {
+          val d0 = (bytes(i) & 0xff) | ((bytes(i + 1) & 0xff) << 8)
+          val d1 = (bytes(i + 2) & 0xff) | ((bytes(i + 3) & 0xff) << 8)
+          val d2 = (bytes(i + 4) & 0xff) | ((bytes(i + 5) & 0xff) << 8)
+
+          i += 6
+
+          val dsum = d0 + d1 + d2
+          if (dsum >= lowerTol) {
+            if (dsum <= upperTol) {
+              val px =
+                if (dsum == 32768)
+                  Array(d0, d1, d2)
+                else
+                  Genotype.weightsToLinear(d0, d1, d2)
+              val gt = Genotype.unboxedGTFromLinear(px)
+              new DosageGenotype(gt, px)
+            } else {
+              setWarning(dosageGreaterThanTolerance)
+              noCall
+            }
+          } else {
+            if (dsum == 0)
+              setWarning(dosageNoCall)
+            else
+              setWarning(dosageLessThanTolerance)
+
+            noCall
+          }
+        }
+      }
+    }
+  }
+}
+
+class BgenRecordV12(compressed: Boolean, nSamples: Int, tolerance: Double) extends BgenRecord {
+  var expectedDataSize: Int = _
+  var expectedNumAlleles: Int = _
+
+  def setExpectedDataSize(size: Int) {
+    this.expectedDataSize = size
+  }
+
+  def setExpectedNumAlleles(n: Int) {
+    this.expectedNumAlleles = n
+  }
+
+  override def getValue: Iterable[Genotype] = {
+    require(input != null, "called getValue before serialized value was set")
+
+    val bytes = if (compressed) decompress(input, expectedDataSize) else input
+    val nRow = (bytes(0) & 0xff) | ((bytes(1) & 0xff) << 8) | ((bytes(2) & 0xff) << 16) | ((bytes(3) & 0xff) << 24)
+    assert(nRow == nSamples, "row nSamples is not equal to header nSamples")
+
+    val nAlleles = (bytes(4) & 0xff) | ((bytes(5) & 0xff) << 8)
+    assert(nAlleles == expectedNumAlleles, s"Value for `nAlleles' in genotype probability data storage is not equal to value in variant identifying data. Expected $expectedNumAlleles but found $nAlleles.")
+
+    val minPloidy = bytes(6) & 0xff
+    val maxPloidy = bytes(7) & 0xff
+
+    if (minPloidy != 2 || maxPloidy != 2)
+      fatal(s"Hail only supports diploid genotypes. Found min ploidy equals `$minPloidy' and max ploidy equals `$maxPloidy'.")
```

Comment:
According to Gavin's wiki page, all genotypes will be diploid. 
https://bitbucket.org/gavinband/bgen/wiki/BGEN_in_the_UK_Biobank